pkg/menus/menus.go

```diff
@@ -0,0 +1,162 @@
+package menus
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/gizak/termui/v3/widgets"
+)
+
+var menuWidth = 50
+
+// Entry contains all the information needed for a boot entry
+type Entry interface {
+	// Label is the string will show in menu
+	Label() string
+	// IsDefault is true means this entry will be hit by default, if there is many default choise, the first on in the list will be choose. the rest will be ignored.
+	IsDefault() bool
+}
+
+// DisplayMenu will present all entries into a menu with numbers, so that user can choose from them.
+func DisplayMenu(menuTitle string, introwords string, location int, entries []Entry) (int, error) {
+	// open a new window
+	if err := ui.Init(); err != nil {
+		log.Printf("failed to initialize termui: %v", err)
+	}
+	// listData contains all choice's labels
+	listData := []string{}
+
+	for i, e := range entries {
+		listData = append(listData, fmt.Sprintf("[%d] %s", i+1, e.Label()))
+	}
+
+	// l is a List item which will present like a menu by termui package
+	l := widgets.NewList()
+	l.Title = menuTitle
+	l.Rows = listData
+	// design the size and location of the menu
+	l.SetRect(0, location, menuWidth, location+len(entries)+2)
+	location += len(entries) + 2
+	l.TextStyle.Fg = ui.ColorWhite
+	ui.Render(l)
+
+	// checkValidFunc is a function for GetInput to check the format of input. Since we want user to choose a option from the menu, the checkValidFunction will check :1.input is a number; 2.input number does not exceed the number of options.
+	checkValidFunc := func(input string) (bool, string) {
+		if input == "" {
+			for _, en := range entries {
+				if en.IsDefault() {
+					return true, ""
+				}
+			}
+			return false, "no default option, please enter a choice"
+		}
+		c, err := strconv.Atoi(input)
+		if err != nil || c < 1 || c > len(entries) {
+			return false, fmt.Sprintf("your input is not a valid entry number.")
+		}
+        return true, ""
+	}
+
+	// call GetInput to get user's choice
+	choose, err := GetInput(introwords, location, 100, 1, checkValidFunc)
+
+	if err != nil {
+		log.Printf("%v", err)
+		ui.Close()
+		return -1, err
+	}
+
+	if choose == "" {
+		for i, en := range entries {
+			if en.IsDefault() {
+				ui.Close()
+				return i, nil
+			}
+		}
+	} else {
+		c, err := strconv.Atoi(choose)
+		ui.Close()
+		return c - 1, err
+	}
+	ui.Close()
+	return -1, nil
+}
+
+// GetInput will present an input box to user and return the user's input.GetInout will check validation of input using checkValidFunc.
+func GetInput(introwords string, location int, inputWidth int, inputHeight int, checkValidFunc func(string) (bool, string)) (string, error) {
```

Comment:
Nit: In Go the convention tends to be to have shorter names, so maybe something like 'desc' instead of introwords and 'isValid' instead of checkValidFunc would be good?

Reply:
In general nice to have shorter variable names!

Reply:
inputWidth -> wid
inputHeight -> ht
and so on